packages/test/test-utils/src/loaderContainerTracker.ts

```diff
@@ -25,36 +54,470 @@ export class LoaderContainerTracker {
+
+        const record = {
+            index: this.containers.size,
+            paused: false,
+            startTrailingNoOps: 0,
+            trailingNoOps: 0,
+            lastProposal: 0,
+        };
+        this.containers.set(container, record);
+        this.trackTrailingNoOps(container, record);
+        this.trackLastProposal(container);
+        this.setupTrace(container, record.index);
+    }
+
+    /**
+     * Keep track of the trailing NoOp that was sent so we can discount them in the clientSequenceNumber tracking.
+     * The server might coalesce them with other ops, or a single NoOp, or delay it if it don't think it is necessary.
+     *
+     * @param container - the container to track
+     * @param record - the record to update the trailing op information
+     */
+    private trackTrailingNoOps(container: IContainer, record: ContainerRecord) {
+        container.deltaManager.outbound.on("op", (messages) => {
+            for (const msg of messages) {
+                if (msg.type === MessageType.NoOp) {
+                    // Track the NoOp that was sent.
+                    if (record.startTrailingNoOps === 0) {
+                        record.startTrailingNoOps = msg.clientSequenceNumber;
+                    }
+                    record.trailingNoOps++;
+                } else {
+                    // Other ops has been sent. We would like to see those ack'ed, so no more need to track NoOps
+                    record.startTrailingNoOps = 0;
+                    record.trailingNoOps = 0;
+                }
+            }
+        });
+
+        container.deltaManager.inbound.on("push", (message) => {
+            // Received the no op back, update the record.
+            if (message.type === MessageType.NoOp
+                && message.clientId === (container as Container).clientId
+                && message.clientSequenceNumber === record.startTrailingNoOps) {
+                record.trailingNoOps--;
+                record.startTrailingNoOps++;
+            }
+        });
+
+        container.on("disconnected", () => {
+            // reset on disconnect.
+            record.startTrailingNoOps = 0;
+            record.trailingNoOps = 0;
+        });
+    }
+
+    private trackLastProposal(container: IContainer) {
+        container.deltaManager.inbound.on("op", (message) => {
+            if (message.type === MessageType.Propose && message.sequenceNumber > this.lastProposalSeqNum) {
+                this.lastProposalSeqNum = message.sequenceNumber;
+            }
+        });
+    }
+
+    /**
+     * Reset the tracker, closing all containers and stop tracking them.
+     */
     public reset() {
-        for (const container of this.containers) {
+        this.lastProposalSeqNum = 0;
+        for (const container of this.containers.keys()) {
             container.close();
         }
         this.containers.clear();
+
+        // REVIEW: do we need to unpatch the loaders?
```

Comment:
probably wise. should be easy if you encapsulate the function you patch with, that function could be stored somewhere on this class, and then you can just shutdown that function somehow, so the loader is still patched, but now the patch does nothing.

Reply:
would also be easier if we had the event, so this could be a follow up

Reply:
Yep... That's why I just left it a comment here, because we might just have the event.